Check generated-visualc-files in check-generated-files

andresag01 opened the pull request:
## Description
Run generated-visualc-files.pl as part of check-generated-files.sh

## Status
**READY**

## Requires Backporting
Yes
mbed TLS 2.1 and 1.3

mpg commented on tests/scripts/check-generated-files.sh:
```diff
@@ -22,3 +22,4 @@ check()
 
 check library/error.c scripts/generate_errors.pl
 check library/version_features.c scripts/generate_features.pl
+check visualc/VS2010/mbedTLS.vcxproj scripts/generate_visualc_files.pl
```
What about the other files generated by this script?

andresag01 replied:
@mpg: Sorry, you are right, I did not look carefully enough. I will modify the script tomorrow to also check the other visual studio files.

hanno-arm commented:
retest

andresag01 commented:
@mpg: I have updated the PR, but I have some reservations about the check-generated-files.sh script.
* I went to look at what files generate-visualc-files.pl generated and it turns out that it generates some general project files for the library, but also some for each app in programs/. The problem is that it did not remove project files that are no longer needed. I found the following project files that were no longer needed and deleted them: `sha1sum` `md5sum` and `sha2sum`. To fix this, I modified the script to remove all the project files and did not check the result of the operation on purpose so that it works even if, for example, the directory is empty.
* I then tried to modify the check-generated-files.sh script. However, I have my reservations about this because generate-visualc-files.pl generates a variable amount of files depending on whats in programs/. As far as I can see we can only check whether there are new files by using git, but I have not done this yet. The modified version of check-generated-files.sh will catch modified and removed files, but not new ones.

What do you think?

mpg commented:
Regarding deleted files, thanks a lot for noticing and fixing this!

Regarding new files, I strongly think that if we have a test, it should be complete in order to avoid a false feeling of confidence. Regarding implementation, I think there are a number of options: we could indeed use git, but I don't think that's the only option: we could just run `ls` before and after and diff the results.

andresag01 commented:
@mpg: I made the script more robust. Now it should catch files created, deleted and modified. Waiting for the CI results.

andresag01 commented:
`mbed TLS build & test` CI has failed the timing test only, the other tests seem to pass.

sbutcher-arm commented on tests/scripts/check-generated-files.sh:
```diff
@@ -11,14 +17,52 @@ fi
+    for FILE in $FILES; do
+        if ! diff $FILE $FILE.bak >/dev/null 2>&1; then
+            echo "'$FILE' was either modified or deleted by '$SCRIPT'"
+            exit 1
```
Exiting here, leaves all the intermediate `*.bak` files behind. Do we really want to do that?

andresag01 replied:
I personally do not mind either way. I left it like this because the behaviour of check-generated-files.sh was already like this anyways. Would you like me to implement this change?

sbutcher-arm replied:
I think it would be better to clear up the files, but if this is what the other scripts do, we've already set a low bar and this is just a test script, mainly for use on the CI which cleans up after itself, so I think your time is better spent elsewhere. Forget it.

sbutcher-arm commented on tests/scripts/check-generated-files.sh:
```diff
@@ -11,14 +17,52 @@ fi
+    # Compare the script output to the old files and remove backups
+    for FILE in $FILES; do
+        if ! diff $FILE $FILE.bak >/dev/null 2>&1; then
+            echo "'$FILE' was either modified or deleted by '$SCRIPT'"
```
If I introduce a new `*.vcxproj` file, I get told it's been modified or deleted, despite it being an addition.

sbutcher-arm replied:
> if you manually add a .vcxproj file and then run check-generated-files.sh the script generate_visualc_files.pl will delete the .vcxproj file if it does not correspond to any app, etc so the change will be reported as a "deleted" file.

That's what I did. I created an artificial invented `.vcproj` file and ran the test. In this case, I guess the script caught the change, and the reported error just confused me, probably because I didn't read it properly. The error is 'x was modified or deleted by script Y', and I guess I misread it as 'x was modified or deleted'. Maybe we could write a better error, but it's only a test script. Forget it.

sbutcher-arm commented on tests/scripts/check-generated-files.sh:
```diff
@@ -11,14 +17,52 @@ fi
+    if [ -d $TO_CHECK ]; then
+        # Check if there are any new files
+        if ls -1 $TO_CHECK | grep -v "$PATTERN" >/dev/null 2>&1; then
+            echo "Files were created by '$SCRIPT'"
```
In my testing, new files were caught at line 41, not here.

andresag01 commented:
@sbutcher-arm: I have applied the requested changes and replied to your comments where appropriate. I have also rebased the PR on top of development. Could you please review the PR once again? Once the changes are approved I will create the backport.

mpg commented on tests/scripts/check-generated-files.sh:
```diff
@@ -1,6 +1,12 @@
-#!/bin/sh
+#! /usr/bin/env sh
```
Not a blocker, just surprised: are there really more systems  with `/usr/bin/env` thand with `/bin/sh`?
(Also, I learned that spaces after `!#` are allowed, which I didn't know.)

sbutcher-arm commented:
Approved, but needs back ports and the CI to pass or be explained.

andresag01 commented:
@sbutcher-arm, @mpg: Thanks for reviewing. I have created the backports:

* Mbed TLS 2.7: https://github.com/ARMmbed/mbedtls/pull/1606
* Mbed TLS 2.1: https://github.com/ARMmbed/mbedtls/pull/1607

sbutcher-arm commented:
Passing CI, back ports reviewed and approved, so now ready for merge.
